File: wasm/src/scene/node.rs
use generational_arena::Index;
use na::{Isometry3, Matrix4, UnitQuaternion, Vector3, Vector4};

#[derive(Debug, Clone)]
pub struct Node {
  pub parent: Option<Index>,
  pub children: Vec<Index>,
  pub matrix_local: Matrix4<f32>,
  pub matrix_world: Matrix4<f32>,
  pub mesh: Option<Index>,
  pub visible: bool,
  pub name: Option<String>,
}

impl Node {
  pub fn new(parent: Option<Index>) -> Self {
    Node {
      parent,
      children: vec![],
      matrix_local: Matrix4::identity(),
      matrix_world: Matrix4::identity(),
      mesh: None,
      visible: true,
      name: None,
    }
  }
}

pub fn compose_matrix(
  position: Option<Vector3<f32>>,
  rotation: Option<UnitQuaternion<f32>>,
  scale: Option<Vector3<f32>>,
) -> Matrix4<f32> {
  let p = position.unwrap_or(Vector3::new(0.0, 0.0, 0.0));
  let r = rotation.unwrap_or(UnitQuaternion::identity());
  let s = scale.unwrap_or(Vector3::new(1.0, 1.0, 1.0));

  let isometry = Isometry3::new(p, r.scaled_axis());
  let node_scale_matrix = Matrix4::from_diagonal(&Vector4::new(s.x, s.y, s.z, 1.0));

  isometry.to_homogeneous() * node_scale_matrix
}

File: src/engine/scene/Node.ts
import { Mesh } from './Mesh';
import { Vector3, Quaternion, Matrix4 } from '../math';

export class Node {
  public readonly mesh?: Mesh;

  public parent?: Node;
  public readonly children: Node[] = [];

  public readonly position = new Vector3();
  public readonly quaternion = new Quaternion();
  public readonly scale = new Vector3(1, 1, 1);

  private readonly matrixLocal = new Matrix4();
  public readonly matrixWorld = new Matrix4();

  public visible: boolean = true;

  add(child: Node): void {
    child.parent = this;
    this.children.push(child);
  }

  remove(child: Node): void {
    child.parent = undefined;
    const index = this.children.indexOf(child);
    if (index !== -1) {
      this.children.splice(index, 1);
    }
  }

  updateMatrixWorld(): void {
    this.matrixLocal.compose(this.position, this.quaternion, this.scale);

    if (this.parent !== undefined) {
      this.matrixWorld.multiplyMatrices(
        this.parent.matrixWorld,
        this.matrixLocal
      );
    } else {
      this.matrixWorld.copy(this.matrixLocal);
    }

    for (let i = 0; i < this.children.length; i++) {
      this.children[i].updateMatrixWorld();
    }
  }

  collectVisibleNodes(): Node[] {
    const nodes: Node[] = [];

    if (this.visible) {
      nodes.push(this);
    }

    for (let i = 0; i < this.children.length; i++) {
      nodes.push(...this.children[i].collectVisibleNodes());
    }

    return nodes;
  }
}
